python/fastfec.py

```diff
@@ -1,59 +1,258 @@
 CUSTOM_WRITE = CFUNCTYPE(None, c_char_p, c_char_p,
                          c_char_p, c_int)
+CUSTOM_LINE = CFUNCTYPE(None, c_char_p, c_char_p, c_char_p)
+
+
+def make_read_buffer(file_input):
+    """Creates a read buffer callback given an open input stream
+
+    Arguments:
+        file_input -- An input stream for the callback to consume
+
+    Returns:
+        A callback function that C code can wrap to consume the input stream"""
 
-def make_read_buffer(f):
     def read_buffer(buffer, want, _):
-        contents_raw = f.read(want)
+        contents_raw = file_input.read(want)
         received = len(contents_raw)
         contents = c_char_p(contents_raw)
         memmove(buffer, contents, received)
         return received
     return read_buffer
 
 
-class FastFEC:
-    def __init__(self, file_handle, output_directory, use_compression=False, make_dirs=False):
-        self.init_lib()
-        self.file_handle = file_handle
-        self.use_compression=use_compression
-        self.make_dirs = make_dirs
+def find_fastfec_lib():
+    """Scans the fastfec shared library in potential locations and returns the path of the found library
+    """
+    prefixes = ["fastfec", "libfastfec"]
+    suffixes = ["so", "dylib", "dll"]
+    directories = [
+        PARENT_DIR,
+        os.path.join(PARENT_DIR, "zig-out/lib"),  # For local dev
+    ]
+
+    # Search in parent directory
+    for root_dir in directories:
+        for prefix in prefixes:
+            for suffix in suffixes:
+                files = glob(os.path.join(root_dir, f"{prefix}*.{suffix}"))
+                if files:
+                    if len(files) > 1:
+                        logging.warn("Expected just one library file")
+                    return os.path.join(PARENT_DIR, files[0])
+
+    # Search in system library directory as last-ditch attempt
+    system_library = find_library("fastfec")
+    if system_library is None:
+        system_library = find_library("libfastfec")
+    return system_library
+
+
+class WriteCache:
+    """Class to store cache information for the custom write function"""
+
+    def __init__(self):
+        self.file_descriptors = {}  # Store all open file descriptors
+        self.last_filename = None  # The last opened filename
+        self.last_fd = None  # The last file descriptor
+
+
+class LineCache:
+    """Class to store cache information for the custom line function"""
+
+    def __init__(self):
+        self.headers = {}  # Store all headers given filename
+        self.last_filename = None  # The last opened filename
+        self.last_headers = None  # The last headers used
+
+
+def csv_parse(line):
+    """Parses a string holding a CSV line into a Python list"""
+    return list(csv.reader([line]))[0]
+
+
+def array_get(array, idx, fallback):
+    """Retrieves the item at position idx in array, or fallback if out of bounds"""
+    try:
+        return array[idx]
+    except IndexError:
+        return fallback
```

Comment:
Are these helpers we can get from some library? They seem pretty simple but if they're necessary here nw